langmuir in "ggl90_mixinglength.F"

hajsong opened the pull request:
## What changes does this PR introduce?
Revised Langmuir circulation (LC) parameterization in a newly written "ggl90_mixinglength.F" file.

## What is the current behaviour? 
The LC parameterization is not revised correctly in "ggl90_mixinglength.F" file, see issue #755.

## What is the new behaviour 
LC parameterization is included in p-coordinate system, too.

## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)


## Other information:

I will be out of touch in a coming 5 days. But I will see the comments when I come back!

## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)
![LC_parameterization](https://github.com/MITgcm/MITgcm/assets/25041806/d3ff702e-9f3f-451c-94df-2399956d65cf)

mjlosch commented:
@hajsong I am afraid that you forgot to check-in / push any changes to `pkg/ggl90/ggl90_mixinglength.F`. This PR contains a modified `data.diagnostics` in `vermi` and a slightly modified `ggl90_mixinglength.F` in `vermix/code` (some comments and new diagnostics).

hajsong commented:
@mjlosch, Thank you for pointing it out. Let me fix that issue. I am sorry for the inconvenience.

mjlosch commented:
@hajsong thanks. The new code looks very simple, which I appreciate. I will try to adjust the AD store directives.

hajsong commented on pkg/ggl90/ggl90_mixinglength.F:
```diff
@@ -294,6 +293,8 @@ SUBROUTINE GGL90_MIXINGLENGTH(
          ENDDO
         ENDDO
        ENDDO
+       CALL DIAGNOSTICS_FILL(GGL90mixingLength,'UDIAG1  ',
+     &                        0,Nr,2,bi,bj,myThid)
```
Yes! I am sorry about that.

hajsong commented on verification/vermix/input.gglLC/data.diagnostics:
```diff
@@ -30,14 +30,19 @@
+#  fields(1:6,3)  = 'GGL90TKE','GGL90Lmx','GGL90Kr ',
+#                   'GGL90ArU','GGL90ArV','GGL90Prl',
+#   fileName(3) = 'DiagMXL_3d',
+#  frequency(3) = -1200.,
 
-  fields(1,4)  = 'MXLDEPTH',
-   fileName(4) = 'DiagMXL_2d',
-  frequency(4) = 432000.,
+#  fields(1,4)  = 'MXLDEPTH',
+#   fileName(4) = 'DiagMXL_2d',
+#  frequency(4) = 432000.,
+
+  fields(1:4,5)  = 'UDIAG1  ','UDIAG2  ','UDIAG3  ','UDIAG4  ',
+   fileName(5) = 'DiagMXL_Dn',
+  frequency(5) = -1200.,
+#  frequency(5) = 432000.,
```
Yes! I should have been more careful for not making unnecessary changes. I am sorry about that.

hajsong commented:
@mjlosch, @jm-c, I am sorry for the late response.  I have been sequestered in a hotel for the last a couple of days, working on problem sets for the university entrance exam. It is pretty strict: no phones or internet allowed. Anyway, I am happy to be back. 
I removed the "UDIAG" line in "ggl90_mixinglength.F" file. Also, I reverted "data.diagnostics" in "input.gglLC".
Please let me know if I am still missing something. Thank you!

mjlosch commented:
@hajsong thanks, looks good to me. Do you have any comments about the new AD-verification experiment? Are there any specific parameter sets that you would like to get tested (to test different stuff than in `vermix.gglLC`)?

mjlosch commented:
@hajsong I noticed that with this branch we have:
```
Y Y Y Y 22 16  3 16> 6<16 16 16  2  0  0 13  1  0  0 16  1 FAIL  vermix.gglLC
```
current master has this:
```
Y Y Y Y 22 16 16 16>16<16 16 16 16 16 16 16 16 16 16 16 16  pass  vermix.gglLC
```
Is that expected?

hajsong commented:
I checked the experiment setting for gglLC and found that the mnc package line was commented out (I think I did that during the test). So I brought that back. Also, I updated the results/output.gglLC.txt. But there are many different lines from the original one because of the different environment of compilation, I think. 
Please let me know if that matters. Thank you!

hajsong commented:
> @hajsong thanks, looks good to me. Do you have any comments about the new AD-verification experiment? Are there any specific parameter sets that you would like to get tested (to test different stuff than in `vermix.gglLC`)?

The current parameter sets are from the paper by Noh et al. (2016) (https://doi.org/10.1175/JPO-D-14-0137.1) which are what I used. I haven't tested other values, but you may try with LC_Gamma=20. This amplifies the mixing length scale even more.

hajsong commented:
Just checking whether you are waiting for me doing something on this...

mjlosch reviewed (approved):
You could add a short suggestion for doc/tag_index in the PR description, otherwise this looks good.

jm-c commented:
I took a quick look and this seems good. Just 2 comments:

1. If it's not too much work, it would be good to know which part of the changes is affecting the `vermix.gglLC` test.
2. @mjlosch regarding the new AD test: I am getting this in `STDERR.0000`:
> (PID.TID 0000.0001) *** ERROR *** GGL90_CHECK: with useCDscheme, missing Coriolis
> (PID.TID 0000.0001) *** ERROR *** GGL90_CHECK:  contrib from Langmuir Parameterization

but the run does not stop (may be this could be changed to a stop ?). Anyway, it might be better to turn off the CD-scheme here. Also since it's a new test, might be better to set `deltaTFreeSurf` to `deltaTtracer` (recommended).

mjlosch commented:
1. I verified that the changes come for the fix in the computation in line 375 (`mxlMaxFlag=3` in `vermix.gglLC`):
```
           IF (GGL90mixingLength(i,j,k) .EQ. mxLength_Dn(i,j,k)) THEN
```
used to be
```
            MLtmp = mxLength_Dn(i,j,k-1)+drF(k-1)*recip_coordFac
            IF (GGL90mixingLength(i,j,k) .EQ. MLtmp) THEN
```
You recover the old results when you revert these lines.

2. I'll make the checks stop the model, and change the verification experiment parameters (and results).

mjlosch commented:
Note: Both `global_oce_latlon.adm.ggl90` and `vermix.gglLC` have a reference output that was not generated on the reference machine.

jm-c commented:
@mjlosch Thanks for pointing to which line of code is responsible for the output difference in `vermix.gglLC` test case. now everything is clear.
@hajsong I looked at code changes, and have few suggestions:

1. I think the case `locMxlMaxFlag=1` is not right for P-coords, because `Ro_surf` is the reference position of the free-surface, which, when using P-coords, is at the bottom ! So I think we still need to distinguish between `usingZcoords` and `usingPcoords` here, and in this case, using "MaxLength= (  rF(k) - R_low(i,j,bi,bj) )*recip_coordFac". For `locMxlMaxFlag` = 2 or 3, there is no problem since `mxLength_Dn` has been set according to P or Z coordinate.
2. with `locMxlMaxFlag` different from 3, the `GGL90_REGULARIZE_MIXINGLENGTH` piece of code (lines 332-333) is going to break the detection of where to increase the `LCmixingLength`. I would suggest to move the full block of code, line 304-342 at the end of this S/R, after the setting of `LCmixingLength`.
3. I don't know if it's a good idea, but regarding the setting of `LCmixingLength`, instead for testing for `GGL90mixingLength = ` something, we could think of testing for `LCmixingLength  >` the same thing. I have the impression that this would give the right answer without beeing prone to some machine truncation issue in the equality test between 2 floating point number.

The last thing is that it would be useful to add the "vermix.gglLC" test to the list of forward test-experiment (in the manual, section 4.15) and also to add "global_oce_latlon.ggl90" test to the list of adjoint test-experiment (in the manual, section 4.16).

mjlosch commented:
> For the second point, `GGL90_REGULARIZE_MIXINGLENGTH` appears in the part related to the `rMixingLength` calculation. This `rMixingLength` is used in the TKE dissipation term, which is not altered by Langmuir circulation effect. So I think it is okay to leave this part as it is.

I have to agree with @jm-c: when I try this, the results of `vermix.gglLC` change. One may ask, why for `locMxlMaxFlag.EQ.3`, `GGL90mixingLength` does not change, but for other values of `locMxlMaxFlag` it does. That can be considered a bug. 
Changing that affects `vermix.gglLC` and probably `global_ocean.oad.ggl90`.

jm-c commented:
@mjlosch What about my suggestion (in point.2)
> to move the full block of code, line 304-342 at the end of this S/R, after the setting of LCmixingLength

mjlosch commented:
I agree, please move the code block. It will require updating the reference.

@hajsong , what do you think about @jm-c's 3rd point?

jm-c commented:
@hajsong I was not very clear indeed. What I was thinking (but not sure it would be doing the right thing) would be to replace, e.g. for case `locMxlMaxFlag` = 2 or 3, line 338
> IF ( GGL90mixingLength(i,j,k) .EQ. mxLength_Dn(i,j,k) ) THEN

with:
> IF ( LCmixingLength(i,j,k) .GE. mxLength_Dn(i,j,k) ) THEN

although I am not sure it should not be ".GT." instead of ".GE.".  
And do a similar change for case `locMxlMaxFlag` = 1 in line 321.

hajsong commented:
@jm-c , Thank you for extra explanation. I am sorry to admit that I am not fully with your suggestion: 
`IF ( LCmixingLength(i,j,k) .GE. mxLength_Dn(i,j,k) ) THEN`.
`LCmixingLength(I,j,k)` was initialized to be a small value (line 140) and not updated up to this point. Perhaps, did you mean 
`IF ( GGL90mixingLength(i,j,k) .GE. mxLength_Dn(i,j,k) ) THEN`?

jm-c commented:
@hajsong I think it's fine to ignore my third point/suggestion for now. 
But just to clarify, since all the different `locMxlMaxFlag` cases only reduce `GGL90mixingLength` ( "GGL90mixingLength(i,j,k) = MIN( GGL90mixingLength(i,j,k) , ...), lines 162-302), I though we could achieve the same result by testing if the original value of `GGL90mixingLength` (stored in `LCmixingLength`) was larger than the limit that comes from the surface BC.

jm-c commented:
I have updated the 2 reference output and also added a Tang-Linear reference output in `global_oce_latlon/results`.
I am going to try to update the list of experiments in the doc.

jm-c commented:
I run some AD testreport (TAF & Tapenade) for experiments that compile pkg/ggl90 and all is good.
So if @hajsong and @mjlosch agree, could merge this PR soon (unless someone wants to add something).

hajsong commented:
@jm-c, I see your point. Since I am currently stuck with my initial approach of how to implement the `LCmixingLength`, I need more time to figure out how to use `GT` or `GE`. I agree with the idea of merging this PR. I will discuss with you when I come up with a new idea. Thank you for your suggestion!

jm-c commented:
@hajsong OK, let's plan to merge this PR as it is now, and if/when a good alternative comes up we could always open a new PR. @mjlosch do you agree ?

mjlosch commented:
I agree!

jm-c commented:
OK, will merge this PR today. Thanks @hajsong and @mjlosch
